core/server/src/main/java/alluxio/master/journal/CheckpointManager.java

```diff
@@ -0,0 +1,145 @@
+   */
+  private final String mTempBackupCheckpointPath;
+  /**
+   * A journal writer through which this checkpoint manager can delete completed logs when the
+   * checkpoint is updated.
+   */
+  private final JournalWriter mWriter;
+
+  /**
+   * Creates a new instance of {@link CheckpointManager}.
+   *
+   * @param ufs the under file system holding the journal
+   * @param the directory for the journal
+   */
+  public CheckpointManager(UnderFileSystem ufs, String checkpointPath, JournalWriter writer) {
+    mUfs = ufs;
+    mCheckpointPath = checkpointPath;
+    mBackupCheckpointPath = mCheckpointPath + ".backup";
+    mTempBackupCheckpointPath = mBackupCheckpointPath + ".tmp";
+    mWriter = writer;
+  }
+
+  /**
+   * Recovers the checkpoint file in case the master crashed while updating it previously.
+   */
+  public void recoverCheckpoint() {
+    try {
+      Preconditions.checkState(
+          !(mUfs.exists(mCheckpointPath) && mUfs.exists(mTempBackupCheckpointPath)
+              && mUfs.exists(mBackupCheckpointPath)),
+          "checkpoint, temp backup checkpoint, and backup checkpoint should never exist "
+              + "simultaneously");
+      if (mUfs.exists(mTempBackupCheckpointPath)) {
+        // If mCheckpointPath exists, step 2 must have implemented rename as copy + delete, and
+        // failed during the delete.
+        UnderFileSystemUtils.deleteIfExists(mUfs, mCheckpointPath);
+        mUfs.rename(mTempBackupCheckpointPath, mCheckpointPath);
+      }
+      if (mUfs.exists(mBackupCheckpointPath)) {
+        // We must have crashed after step 3
+        if (mUfs.exists(mCheckpointPath)) {
+          // We crashed after step 4, so we can finish steps 5 and 6.
+          mWriter.deleteCompletedLogs();
+          mUfs.delete(mBackupCheckpointPath, false);
+        } else {
+          // We crashed before step 4, so we roll back to backup checkpoint.
+          mUfs.rename(mBackupCheckpointPath, mCheckpointPath);
+        }
+      }
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Updates the checkpoint file to the checkpoint at the specified path. The update is done in such
+   * a way that {@link #recoverCheckpoint()} will recover from any failures that occur during the
+   * update.
+   *
+   * @param newCheckpointPath the path to the new checkpoint file
+   */
+  public void updateCheckpoint(String newCheckpointPath) {
+    try {
+      if (mUfs.exists(mCheckpointPath)) {
+        UnderFileSystemUtils.deleteIfExists(mUfs, mTempBackupCheckpointPath);
+        UnderFileSystemUtils.deleteIfExists(mUfs, mBackupCheckpointPath);
+        // Rename in two steps so that we never have identical mCheckpointPath and
+        // mBackupCheckpointPath. This is a concern since UFS may implement rename as copy + delete.
+        mUfs.rename(mCheckpointPath, mTempBackupCheckpointPath);
+        mUfs.rename(mTempBackupCheckpointPath, mBackupCheckpointPath);
```

Comment:
NIT: maybe add a log message after this that the backup of the checkpoint was completed?